File: backend/app/api/articles.py
import json
import math
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func, or_
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import Article, Source
from app.schemas import ArticleOut, ArticleDetail, ArticleListResponse

router = APIRouter()


@router.get("/articles", response_model=ArticleListResponse)
async def list_articles(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    company: Optional[str] = None,
    topic: Optional[str] = None,
    q: Optional[str] = None,
    sort: str = Query("newest", pattern="^(newest|oldest)$"),
    db: AsyncSession = Depends(get_db),
):
    base_query = (
        select(Article, Source.company, Source.name)
        .join(Source, Article.source_id == Source.id)
        .where(Article.status == "done")
    )

    if company:
        base_query = base_query.where(Source.company == company)
    if topic:
        base_query = base_query.where(Article.topics.contains(topic.lower()))
    if q:
        q_lower = f"%{q.lower()}%"
        base_query = base_query.where(
            or_(
                func.lower(Article.title).like(q_lower),
                func.lower(Article.summary).like(q_lower),
            )
        )

    # Count total
    count_query = select(func.count()).select_from(base_query.subquery())
    total = await db.scalar(count_query) or 0

    # Ordering
    order_col = Article.published_at if sort in ("newest", "oldest") else Article.published_at
    if sort == "newest":
        base_query = base_query.order_by(order_col.desc().nullslast())
    else:
        base_query = base_query.order_by(order_col.asc().nullsfirst())

    # Pagination
    offset = (page - 1) * limit
    base_query = base_query.offset(offset).limit(limit)

    result = await db.execute(base_query)
    rows = result.all()

    items = []
    for article, company_name, source_name in rows:
        item = ArticleOut(
            id=article.id,
            source_id=article.source_id,
            url=article.url,
            title=article.title,
            author=article.author,
            published_at=article.published_at,
            crawled_at=article.crawled_at,
            summary=article.summary,
            summarized_at=article.summarized_at,
            topics=article.topics,
            is_relevant=article.is_relevant,
            status=article.status,
            word_count=article.word_count,
            company=company_name,
            source_name=source_name,
        )
        items.append(item)

    # Collect distinct companies and topics for filter UI
    companies_result = await db.execute(
        select(Source.company)
        .join(Article, Article.source_id == Source.id)
        .where(Article.status == "done")
        .distinct()
        .order_by(Source.company)
    )
    companies = [r[0] for r in companies_result.all()]

    # Aggregate all topics from done articles
    topics_result = await db.execute(
        select(Article.topics).where(Article.status == "done").where(Article.topics != "[]")
    )
    topic_counter: dict[str, int] = {}
    for (topics_json,) in topics_result.all():
        try:
            for t in json.loads(topics_json):
                topic_counter[t] = topic_counter.get(t, 0) + 1
        except Exception:
            pass
    top_topics = sorted(topic_counter, key=lambda x: -topic_counter[x])[:30]

    return ArticleListResponse(
        items=items,
        total=total,
        page=page,
        pages=max(1, math.ceil(total / limit)),
        companies=companies,
        topics=top_topics,
    )


@router.get("/articles/{article_id}", response_model=ArticleDetail)
async def get_article(article_id: int, db: AsyncSession = Depends(get_db)):
    from fastapi import HTTPException

    result = await db.execute(
        select(Article, Source.company, Source.name)
        .join(Source, Article.source_id == Source.id)
        .where(Article.id == article_id)
    )
    row = result.first()
    if not row:
        raise HTTPException(status_code=404, detail="Article not found")

    article, company_name, source_name = row
    return ArticleDetail(
        id=article.id,
        source_id=article.source_id,
        url=article.url,
        title=article.title,
        author=article.author,
        published_at=article.published_at,
        crawled_at=article.crawled_at,
        summary=article.summary,
        summarized_at=article.summarized_at,
        topics=article.topics,
        is_relevant=article.is_relevant,
        status=article.status,
        word_count=article.word_count,
        company=company_name,
        source_name=source_name,
        raw_content=article.raw_content,
        summary_model=article.summary_model,
        relevance_score=article.relevance_score,
        error_message=article.error_message,
    )

File: frontend/src/api/articles.ts
export interface Article {
  id: number
  url: string
  title: string
  author: string | null
  published_at: string | null
  crawled_at: string
  summary: string | null
  topics: string // JSON string
  word_count: number | null
  company: string | null
  source_name: string | null
}

export interface ArticleListResponse {
  items: Article[]
  total: number
  page: number
  pages: number
  companies: string[]
  topics: string[]
  updatedAt: string | null
}

export interface ArticleFilters {
  page?: number
  limit?: number
  company?: string
  topic?: string
  q?: string
  sort?: 'newest' | 'oldest'
}

// Cache the raw articles so we only fetch once per session
let _cache: Article[] | null = null

async function loadAll(): Promise<Article[]> {
  if (_cache) return _cache
  const res = await fetch('./articles.json')
  if (!res.ok) throw new Error(`Failed to load articles.json: HTTP ${res.status}`)
  _cache = await res.json()
  return _cache!
}

export async function fetchArticles(filters: ArticleFilters = {}): Promise<ArticleListResponse> {
  const {
    page = 1,
    limit = 20,
    company,
    topic,
    q,
    sort = 'newest',
  } = filters

  let all = await loadAll()

  // Filter
  if (company) all = all.filter(a => a.company === company)
  if (topic) all = all.filter(a => a.topics.toLowerCase().includes(topic.toLowerCase()))
  if (q) {
    const ql = q.toLowerCase()
    all = all.filter(a =>
      a.title.toLowerCase().includes(ql) ||
      (a.summary ?? '').toLowerCase().includes(ql)
    )
  }

  // Sort
  all = [...all].sort((a, b) => {
    const da = a.published_at ? new Date(a.published_at).getTime() : 0
    const db = b.published_at ? new Date(b.published_at).getTime() : 0
    return sort === 'newest' ? db - da : da - db
  })

  const total = all.length
  const pages = Math.max(1, Math.ceil(total / limit))
  const offset = (page - 1) * limit
  const items = all.slice(offset, offset + limit)

  // Build filter options from full unfiltered dataset
  const everything = await loadAll()
  const companySet = new Set(everything.map(a => a.company).filter(Boolean) as string[])

  const topicCounter: Record<string, number> = {}
  for (const a of everything) {
    try {
      for (const t of JSON.parse(a.topics)) {
        topicCounter[t] = (topicCounter[t] ?? 0) + 1
      }
    } catch {}
  }
  const topTopics = Object.keys(topicCounter).sort((a, b) => topicCounter[b] - topicCounter[a]).slice(0, 30)

  // Try to get last updated time from newest article
  const latest = everything[0]?.crawled_at ?? null

  return {
    items,
    total,
    page,
    pages,
    companies: [...companySet].sort(),
    topics: topTopics,
    updatedAt: latest,
  }
}

export function parseTopics(topicsJson: string): string[] {
  try {
    return JSON.parse(topicsJson)
  } catch {
    return []
  }
}
